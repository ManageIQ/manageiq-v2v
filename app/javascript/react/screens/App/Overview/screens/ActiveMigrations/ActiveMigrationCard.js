import _ from 'lodash';
import moment from 'moment';
import React from 'react';
import PropTypes from 'prop-types';
import {
  Grid,
  Icon,
  OverlayTrigger,
  Tooltip,
  UtilizationBar
} from 'patternfly-react';

const ActiveMigrationCard = ({ migration }) => {
  const availableTooltip = (max, now) => {
    if (max > 0) {
      return (
        <Tooltip id={`available-tip-${migration.id}`}>
          {sprintf(__('%s%% Remaining'), Math.round((max - now) / max * 100))}
        </Tooltip>
      );
    }
    return (
      <Tooltip id={`available-tip-${migration.id}`}>{__('No Data')}</Tooltip>
    );
  };
  const usedTooltip = (max, now) => {
    if (max > 0) {
      return (
        <Tooltip id={`available-tip-${migration.id}`}>
          {sprintf(__('%s%% Complete'), Math.round(now / max * 100))}
        </Tooltip>
      );
    }
    return (
      <Tooltip id={`available-tip-${migration.id}`}>{__('No Data')}</Tooltip>
    );
  };

  const totalVMs = _.get(migration, 'miq_request_tasks', []).length;
  const completedVMs = _.filter(_.get(migration, 'miq_request_tasks', []), {
    state: 'finished'
  }).length;

  const startTime = new Date(_.get(migration, 'options.delivered_on'));
  const endTime = Date.now();
  const elapsedHours = moment(endTime).diff(startTime, 'hours');
  const elapsedMinutes = moment(endTime).diff(startTime, 'minutes') % 60;

  let elapsedTime;
  if (elapsedHours >= 48) {
    elapsedTime = sprintf(
      __('%s days %s:%s elapsed'),
      Math.floor(elapsedHours / 24),
      elapsedHours % 24,
      elapsedMinutes
    );
  } else if (elapsedHours >= 24) {
    elapsedTime = sprintf(
      __('1 day %s:%s elapsed'),
      elapsedHours % 24,
      elapsedMinutes
    );
  } else {
    elapsedTime = sprintf(__('%s:%s elapsed'), elapsedHours, elapsedMinutes);
  }

  const vmBarLabel = (
    <span>
      <strong className="label-strong">
        {sprintf(__('%s of %s VMs'), completedVMs, totalVMs)}
      </strong>{' '}
      {__('Migrated')}
    </span>
  );
  const usedVmTooltip = () => usedTooltip(totalVMs, completedVMs);
  const availableVmTooltip = () => availableTooltip(totalVMs, completedVMs);

  return (
    <Grid.Col sm={12} md={6} lg={4}>
      <div className="card-pf">
        <div className="card-pf-heading">
          <OverlayTrigger
            overlay={
              <Tooltip id={`description_${migration.id}`}>
                {migration.description}
              </Tooltip>
            }
            placement="top"
            trigger={['hover']}
            delay={500}
            rootClose={false}
          >
            <h3>{migration.description}</h3>
          </OverlayTrigger>
        </div>
        <div className="card-pf-body">
          <UtilizationBar
            now={completedVMs || 1}
            max={totalVMs}
            description={__('VMs')}
            label={vmBarLabel}
            descriptionPlacementTop
            availableTooltipFunction={availableVmTooltip}
            usedTooltipFunction={usedVmTooltip}
          />
          <div className="active-migration-elapsed-time">
            <Icon type="fa" name="clock-o" />
            {elapsedTime}
          </div>
        </div>
      </div>
    </Grid.Col>
  );
};

ActiveMigrationCard.propTypes = {
  migration: PropTypes.object
};
ActiveMigrationCard.defaultProps = {
  migration: {}
};

export default ActiveMigrationCard;
