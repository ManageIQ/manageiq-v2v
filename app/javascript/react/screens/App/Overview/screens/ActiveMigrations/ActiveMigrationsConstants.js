export const FETCH_V2V_ACTIVE_MIGRATIONS = 'FETCH_V2V_ACTIVE_MIGRATIONS';
