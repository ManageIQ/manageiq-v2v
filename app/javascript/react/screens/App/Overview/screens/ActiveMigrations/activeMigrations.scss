.card-pf-multi-card-container {
  background: transparent;
  border: none;
  box-shadow: none;
  .card-pf-heading h3 {
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }
  .card-pf-body > .message-text {
    font-size: 16px;
    margin-left: 5px;
  }
}

.active-migration-elapsed-time {
  .fa,
  .pficon {
    margin-right: 5px;
  }
}
