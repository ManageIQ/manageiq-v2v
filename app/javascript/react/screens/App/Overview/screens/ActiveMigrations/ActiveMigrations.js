import React from 'react';
import PropTypes from 'prop-types';
import { Grid, Icon, Spinner } from 'patternfly-react';
import ActiveMigrationCard from './ActiveMigrationCard';

class ActiveMigrations extends React.Component {
  componentDidMount() {
    const { fetchActiveMigrationsAction } = this.props;
    fetchActiveMigrationsAction();
  }

  renderActiveMigrations() {
    const { activeMigrations } = this.props;

    return (
      <Grid fluid>
        <Grid.Row>
          {activeMigrations.map(migration => (
            <ActiveMigrationCard migration={migration} key={migration.id} />
          ))}
        </Grid.Row>
      </Grid>
    );
  }

  render() {
    const {
      isRejectedActiveMigrations,
      isFetchingActiveMigrations,
      errorActiveMigrations,
      activeMigrations
    } = this.props;

    return (
      <div className="card-pf card-pf-multi-card-container">
        <div className="card-pf-heading">
          <h2>
            {sprintf(__('%s Migrations in Progress'), activeMigrations.length)}
          </h2>
        </div>
        <div className="card-pf-body">
          {isFetchingActiveMigrations && (
            <React.Fragment>
              <Spinner size="xs" inline loading />
              <span className="message-text">
                {__('Loading migrations in progress')}
              </span>
            </React.Fragment>
          )}
          {isRejectedActiveMigrations && (
            <React.Fragment>
              <Icon type="pf" name="error-circle-o" />
              <span className="message-text">
                {errorActiveMigrations ||
                  __('Error retrieving active migrations')}
              </span>
            </React.Fragment>
          )}
          {activeMigrations.length > 0 && this.renderActiveMigrations()}
        </div>
      </div>
    );
  }
}

ActiveMigrations.propTypes = {
  fetchActiveMigrationsAction: PropTypes.func,
  isFetchingActiveMigrations: PropTypes.bool,
  activeMigrations: PropTypes.arrayOf(PropTypes.object),
  isRejectedActiveMigrations: PropTypes.bool,
  errorActiveMigrations: PropTypes.string
};
ActiveMigrations.defaultProps = {
  activeMigrations: [],
  isFetchingActiveMigrations: false,
  isRejectedActiveMigrations: false,
  errorActiveMigrations: ''
};

export default ActiveMigrations;
