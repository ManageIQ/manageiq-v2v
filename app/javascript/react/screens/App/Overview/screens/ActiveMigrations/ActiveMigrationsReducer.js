import Immutable from 'seamless-immutable';
import _ from 'lodash';

import { FETCH_V2V_ACTIVE_MIGRATIONS } from './ActiveMigrationsConstants';

const initialState = Immutable({
  activeMigrations: [],
  isFetchingActiveMigrations: false,
  isRejectedActiveMigrations: false,
  errorActiveMigrations: ''
});

export default (state = initialState, action) => {
  switch (action.type) {
    case `${FETCH_V2V_ACTIVE_MIGRATIONS}_PENDING`:
      return state.set('isFetchingActiveMigrations', true);
    case `${FETCH_V2V_ACTIVE_MIGRATIONS}_FULFILLED`:
      return state
        .set('activeMigrations', _.get(action, 'payload.data.resources', []))
        .set('isFetchingActiveMigrations', false);
    case `${FETCH_V2V_ACTIVE_MIGRATIONS}_REJECTED`:
      return state
        .set('errorActiveMigrations', action.payload)
        .set('isRejectedActiveMigrations', true)
        .set('isFetchingActiveMigrations', false);
    default:
      return state;
  }
};
