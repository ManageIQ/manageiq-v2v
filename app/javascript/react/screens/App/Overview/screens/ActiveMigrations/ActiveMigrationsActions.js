import URI from 'urijs';
import API from '../../../../../../common/API';

import { FETCH_V2V_ACTIVE_MIGRATIONS } from './ActiveMigrationsConstants';
import { requestActiveServiceRequests } from './activeMigrations.fixtures';

const mockMode = true;

const _getActiveMigrationsAction = url => dispatch => {
  if (mockMode) {
    dispatch({
      type: `${FETCH_V2V_ACTIVE_MIGRATIONS}_FULFILLED`,
      payload: requestActiveServiceRequests.response
    });
    return;
  }
  dispatch({
    type: FETCH_V2V_ACTIVE_MIGRATIONS,
    payload: API.get(url)
  });
};

const fetchActiveMigrationsAction = () => {
  const url =
    '/api/service_requests?' +
    "filter[]=state='active'" +
    "&filter[]=type='ServiceTemplateMigrationPlanRequest'" +
    '&expand=resources' +
    '&attributes=status,state,options,description,miq_request_tasks';

  const uri = new URI(url);

  return _getActiveMigrationsAction(uri.toString());
};

export { fetchActiveMigrationsAction };
