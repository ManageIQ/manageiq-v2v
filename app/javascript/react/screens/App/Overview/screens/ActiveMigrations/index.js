import { connect } from 'react-redux';

import ActiveMigrations from './ActiveMigrations';
import reducer from './ActiveMigrationsReducer';
import * as MigrationsActions from './ActiveMigrationsActions';

export const reducers = { activeMigrations: reducer };

const mapStateToProps = ({ activeMigrations }, ownProps) => ({
  ...activeMigrations,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(mapStateToProps, MigrationsActions, mergeProps)(
  ActiveMigrations
);
