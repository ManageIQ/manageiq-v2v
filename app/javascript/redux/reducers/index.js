import { combineReducers } from 'redux';
import { reducer as formReducer } from 'redux-form';
import { reducers as overviewReducers } from '../../react/screens/App/Overview';
import { reducers as mappingWizardReducers } from '../../react/screens/App/Overview/screens/MappingWizard';
import { reducers as mappingWizardClustersStepReducers } from '../../react/screens/App/Overview/screens/MappingWizard/components/MappingWizardClustersStep';
import { reducers as mappingWizardDatastoresStepReducers } from '../../react/screens/App/Overview/screens/MappingWizard/components/MappingWizardDatastoresStep';
import { reducers as mappingWizardNetworksStepReducers } from '../../react/screens/App/Overview/screens/MappingWizard/components/MappingWizardNetworksStep';
import { reducers as mappingWizardResultsStepReducers } from '../../react/screens/App/Overview/screens/MappingWizard/components/MappingWizardResultsStep';
import { reducers as planWizardReducers } from '../../react/screens/App/Overview/screens/PlanWizard';
import { reducers as planWizardResultsStepReducers } from '../../react/screens/App/Overview/screens/PlanWizard/components/PlanWizardResultsStep';

export function combineReducersAsync(asyncReducers) {
  return combineReducers({
    ...overviewReducers,
    ...mappingWizardReducers,
    ...mappingWizardClustersStepReducers,
    ...mappingWizardDatastoresStepReducers,
    ...mappingWizardNetworksStepReducers,
    ...mappingWizardResultsStepReducers,
    ...planWizardReducers,
    ...planWizardResultsStepReducers,
    form: formReducer,
    ...asyncReducers
  });
}

export default combineReducersAsync();
