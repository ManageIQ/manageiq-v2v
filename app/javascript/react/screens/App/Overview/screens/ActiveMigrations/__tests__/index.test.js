import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { initialState } from '../activeMigrations.fixtures';
import ActiveMigrations from '../index';
import reducer from '../ActiveMigrationsReducer';

const reducers = { activeMigrations: reducer };

describe('Active ActiveMigrations integration test', () => {
  const middlewares = [thunk, promiseMiddleware()];
  const generateStore = () =>
    createStore(
      combineReducers({ ...reducers }),
      {
        activeMigrations: initialState
      },
      applyMiddleware(...middlewares)
    );

  const mountComponent = store =>
    mount(
      <Provider store={store}>
        <ActiveMigrations />
      </Provider>
    );

  it('should mount the ActiveMigrations with mapStateToProps reduced', () => {
    const store = generateStore();
    const wrapper = mountComponent(store);

    // query the unconnected component to assert reduced props
    const component = wrapper.find(ActiveMigrations);
    expect(component.props()).toMatchSnapshot();
  });
});
