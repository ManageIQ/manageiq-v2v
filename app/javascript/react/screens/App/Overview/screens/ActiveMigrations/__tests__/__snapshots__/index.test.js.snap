// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active ActiveMigrations integration test should mount the ActiveMigrations with mapStateToProps reduced 1`] = `Object {}`;
